README
The following library contain source code for validating the checksum and structure of known credit cards companies.
This information is released under LGPL and intended to help creating free library for checking validity of credit cards.
The intent if the long run of this library it to be a shard library so all programming languages can use it inside.
The code does not arrive instead of automatic banking validation systems, but as a measure that should arrive prior of using such system.

The following version 0.1 beta is not fully checked, and there are few known problems:
  1. Not a full support for JCB - There is a need for additional information regarding that credit card.
  2. Not a full support for Isracard - There is a need for additional information regarding the allowed prefix of the credit card, and a generator to validate numbers as much as possible.
  3. There are several credit cards that I did not find any generators for, so I did not check them.  
  4. The current list of supported credit cards (including the ones that where not checked yet):
     * MasterCard
     * Visa (13 and 16 in length)
     * Visa Electron
     * American Express
     * Diners Club Carte Blanche
     * Diners Club International
     * Diners Club US and Canada
     * Discover
     * JCB
     * Laser
     * Maestro
     * Solo
     * Switch
     * Isracard

If you have information about additional credit cards or you can help working on existed credit cards, please donate your code to this units so anyone could use your work.

Files included by this package:
  * README                     - This file
  * lgpl-3.0.txt               - The library license
  * src/test.pp                - Some simple tests for the library (work in progress)
  * src/luhn.pp                - Implementation of the Luhn algorithm 
  * src/creditcard_validity.pp - The validation library (contain documentation inside for now -> will be extracted on next versions to an html document using fpdoc)
  
=============================================================================================
= This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; =
= without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. =  
= See the GNU Lesser General Public License for more details.                               =
=                                                                                           =
=    IF YOU DO NOT AGREE WITH THE TERMS OF THE LGPL LICENSE, DO NOT USE THIS LIBRARY !!!    =
=============================================================================================



releases
../../release/creditcard/

src/tests/testcases/gencc.php
<?php
/*
PHP credit card number generator
Copyright (C) 2006 Graham King graham@darkcoding.net

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

$visaPrefixList[] =  "4539";
$visaPrefixList[] =  "4556";
$visaPrefixList[] =  "4916";
$visaPrefixList[] =  "4532";
$visaPrefixList[] =  "4929";
$visaPrefixList[] =  "40240071";
$visaPrefixList[] =  "4485";
$visaPrefixList[] =  "4716";
$visaPrefixList[] =  "4";

$mastercardPrefixList[] =  "51";
$mastercardPrefixList[] =  "52";
$mastercardPrefixList[] =  "53";
$mastercardPrefixList[] =  "54";
$mastercardPrefixList[] =  "55";

$amexPrefixList[] =  "34";
$amexPrefixList[] =  "37";

$discoverPrefixList[] = "6011";

$dinersPrefixList[] =  "300";
$dinersPrefixList[] =  "301";
$dinersPrefixList[] =  "302";
$dinersPrefixList[] =  "303";
$dinersPrefixList[] =  "36";
$dinersPrefixList[] =  "38";

$enRoutePrefixList[] =  "2014";
$enRoutePrefixList[] =  "2149";

$jcbPrefixList16[] =  "3088";
$jcbPrefixList16[] =  "3096";
$jcbPrefixList16[] =  "3112";
$jcbPrefixList16[] =  "3158";
$jcbPrefixList16[] =  "3337";
$jcbPrefixList16[] =  "3528";

$jcbPrefixList15[] = "2100";
$jcbPrefixList15[] = "1800";

$voyagerPrefixList[] = "8699";

/*
'prefix' is the start of the CC number as a string, any number of digits.
'length' is the length of the CC number to generate. Typically 13 or 16
*/
function completed_number($prefix, $length) {

    $ccnumber = $prefix;

    # generate digits

    while ( strlen($ccnumber) < ($length - 1) ) {
        $ccnumber .= rand(0,9);
    }

    # Calculate sum

    $sum = 0;
    $pos = 0;

    $reversedCCnumber = strrev( $ccnumber );

    while ( $pos < $length - 1 ) {

        $odd = $reversedCCnumber[ $pos ] * 2;
        if ( $odd > 9 ) {
            $odd -= 9;
        }

        $sum += $odd;

        if ( $pos != ($length - 2) ) {

            $sum += $reversedCCnumber[ $pos +1 ];
        }
        $pos += 2;
    }

    # Calculate check digit

    $checkdigit = (( floor($sum/10) + 1) * 10 - $sum) % 10;
    $ccnumber .= $checkdigit;

    return $ccnumber;
}

function credit_card_number($prefixList, $length, $howMany) {

    for ($i = 0; $i < $howMany; $i++) {

        $ccnumber = $prefixList[ array_rand($prefixList) ];
        $result[] = completed_number($ccnumber, $length);
    }

    return $result;
}

function output($title, $numbers) {

    $result[] = "<div class='creditCardNumbers'>";
    $result[] = "<h3>$title</h3>";
    $result[] = implode('<br />', $numbers);
    $result[]= '</div>';

    return implode('<br />', $result);
}

#
# Main
#

echo "<div class='creditCardSet'>";
$mastercard = credit_card_number($mastercardPrefixList, 16, 10);
echo output("Mastercard", $mastercard);

$visa16 = credit_card_number($visaPrefixList, 16, 10);
echo output("VISA 16 digit", $visa16);
echo "</div>";

echo "<div class='creditCardSet'>";
$visa13 = credit_card_number($visaPrefixList, 13, 5);
echo output("VISA 13 digit", $visa13);

$amex = credit_card_number($amexPrefixList, 15, 5);
echo output("American Express", $amex);
echo "</div>";

# Minor cards

echo "<div class='creditCardSet'>";
$discover = credit_card_number($discoverPrefixList, 16, 3);
echo output("Discover", $discover);

$diners = credit_card_number($dinersPrefixList, 14, 3);
echo output("Diners Club", $diners);
echo "</div>";

echo "<div class='creditCardSet'>";
$enRoute = credit_card_number($enRoutePrefixList, 15, 3);
echo output("enRoute", $enRoute);

$jcb15 = credit_card_number($jcbPrefixList15, 15, 3);
echo output("JCB 15 digit", $jcb15);
echo "</div>";

echo "<div class='creditCardSet'>";
$jcb16 = credit_card_number($jcbPrefixList16, 16, 3);
echo output("JCB 16 digit", $jcb16);

$voyager = credit_card_number($voyagerPrefixList, 15, 3);
echo output("Voyager", $voyager);
echo "</div>";
?>


src/tests/testcases/ccvalidate.js
/*
 * Script Name: ccvalidate.js
 * Written By:  Kevin King                    (Kevin@PrecisOnline.com)
 *              for Precision Solutions, Inc. (http://www.PrecisOnline.com)
 *
 * Date:        28 Oct 2002
 *
 * Description: This script validates the basic accuracy of a credit card 
 *              number.  It expects the card type and card number to have 
 *              been entered/selected, and then the verifyCard function can 
 *              be called to verify the accuracy of the card number.
 *
 *              This script does not ensure that the card is active, so it 
 *              is possible that while the card number is accurate, the card 
 *              itself may have been deactivated by the issuer (i.e. for lost 
 *              or stolen cards).
 *
 *              THIS SOFTWARE IS PROVIDED AS-IS AND AUTHOR PROVIDES NO 
 *              WARRANTIES, EITHER EXPRESSED OR IMPLIED, REGARDING THE 
 *              APPLICABILITY OR FITNESS OF THIS SOFTWARE FOR ANY SPECIFIC 
 *              PURPOSE.
 *
 *              By using this software, you implicitly warrant your use
 *              will conform to normal legitimate business purposes.  Author 
 *              is not responsible for, and will assist authorities in the 
 *              prosecution of, any use of this software that is in violation
 *              of United States or international law.
 *
 *****************************************************************************
 * Modifications
 *****************************************************************************
 * Date.....  Changed By.....  Description of Change..........................
 * 30 Jan 06  Kevin King       Added Australian BankCard support per details
 *                             received from Paul Bronshteyn (thanks again,
 *                             Paul!).
 * 28 Jan 06  Kevin King       Corrected Delta rules as identified by Paul
 *                             Bronshteyn (thanks Paul!).
 * 26 Jan 06  Kevin King       Added checkCvc function to verify CVV/CID values.
 * 20 Jan 06  Kevin King       Updated Maestro and Solo rules per
 *                             http://www.barclaycardmerchantservices.co.uk/existing_customers/operational/pdf/binranges.pdf
 * 16 Mar 05  Kevin King       Updated to add Delta and Maestro card support.
 * 05 Apr 03  Kevin King       Updated SWITCH card validation table.  Added
 *                             support for validating SOLO cards.
 */

var cc_checksum;                   // Number: The calculated checksum for the card
var cc_fail_reason;                // Text: Explains the reason for the failure
var cc_switch_startDateLength;     // Number: Count of digits in start date or 0=no start date  (Switch/SOLO)
var cc_switch_issueLength;         // Number: Count of digits in issue number or 0=no issue number (Switch/SOLO)
var cc_cvc_length;                 // Number: Count of digits in CVV2

/*
 *****************************************************************************
 * Verify the check digit
 *****************************************************************************
 */
  
function luhnCheck(cardNbr)
{
    var result     = true;
  
    var ndx        = cardNbr.length - 1;
    var checkDigit = cardNbr.substr(ndx,1);
    var multiplier = 2;
    var accum      = 0;
  
    while(ndx)
    {
        var cardCh = cardNbr.substr(ndx - 1,1);
        if(cardCh >= 0 && cardCh <= 9)
        {
            var thisValue  = cardCh * multiplier;
            while(thisValue > 0)
            {
                accum += Math.floor(thisValue % 10);
                thisValue = Math.floor(thisValue / 10);
            }

            multiplier = (multiplier == 2 ? 1 : 2);
        }

        ndx--;
    }
  
    cc_checksum = ((10 - (accum % 10)) % 10);
    if(cc_checksum != checkDigit)
        result = false;

    return(result);
}

/*
 *****************************************************************************
 * Verify the card number is all numbers
 *****************************************************************************
 */

function checkDigits(cardNbr)
{
    var result = true;
    var ndx;
  
    for(ndx = 0 ; ndx < cardNbr.length && result ; ++ndx)
    {
        var cardCh = cardNbr.substr(ndx,1);
        if(cardCh < '0' || cardCh > '9')
            result = false;
    }
  
    return(result);
}

/*
 *****************************************************************************
 * Verify the prefix and length is proper for this type of card
 *****************************************************************************
 */
 
function checkPrefixAndLength(cardType, cardNbr)
{
    var result    = false;
  
    var amexRules    = new Array("37,37,15,3","34,34,15,3");
    var discRules    = new Array("6011,6011,16");
    var dinersRules  = new Array("36,36,14","38,38,14","300,305,14");
    var enrouteRules = new Array("2014,2014,15","2149,2149,15");
    var jcbRules     = new Array("3,3,16","2131,2131,15","1800,1800,15");
    var mcRules      = new Array("51,55,16,3");
    var visaRules    = new Array("4,4,13,3","4,4,16,3");
    var switchRules  = new Array("490302,490309,18,,1","490335,490339,18,,1","491101,491102,16,,1",
                                 "491174,491182,18,,1","493600,493699,19,,1","564182,564182,16,,2",
                                 "633300,633300,16,,0","633301,633301,19,,1","633302,633349,16,,0",
                                 "675900,675900,16,,0","675901,675901,19,,1","675902,675904,16,,0",
                                 "675905,675905,19,,1","675906,675917,16,,0","675918,675918,19,,1",
                                 "675919,675937,16,,0","675938,675940,18,,1","675941,675949,16,,0",
                                 "675950,675962,19,,1","675963,675997,16,,0","675998,675998,19,,1",
                                 "675999,675999,16,,0");
    var soloRules    = new Array("633450,633453,16,,0","633454,633457,16,,0","633458,633460,16,,0",
                                 "633461,633461,18,,1","633462,633472,16,,0","633473,633473,18,,1",
                                 "633474,633475,16,,0","633476,633476,19,,1","633477,633477,16,,0",
                                 "633478,633478,18,,1","633479,633480,16,,0","633481,633481,19,,1",
                                 "633482,633489,16,,0","633490,633493,16,,1","633494,633494,18,,1",
                                 "633495,633497,16,,2","633498,633498,19,,1","633499,633499,18,,1",
                                 "676700,676700,16,,0","676701,676701,19,,1","676702,676702,16,,0",
                                 "676703,676703,18,,1","676704,676704,16,,0","676705,676705,19,,1",
                                 "676706,676707,16,,2","676708,676711,16,,0","676712,676715,16,,0",
                                 "676716,676717,16,,0","676718,676718,19,,1","676719,676739,16,,0",
                                 "676740,676740,18,,1","676741,676749,16,,0","676750,676762,19,,1",
                                 "676763,676769,16,,0","676770,676770,19,,1","676771,676773,16,,0",
                                 "676774,676774,18,,1","676775,676778,16,,0","676779,676779,18,,1",
                                 "676780,676781,16,,0","676782,676782,18,,1","676783,676794,16,,0",
                                 "676795,676795,18,,1","676796,676797,16,,0","676798,676798,19,,1",
                                 "676799,676799,16,,0");
    var maestroRules = new Array("490303,490303,16/18/19","493698,493699,16/18/19","633302,633349,16/18/19",
                                 "675900,675999,16/18/19");
    var deltaRules   = new Array("413733,413737,16","446200,446299,16","453978,453979,16",
                                 "454313,454313,16","454432,454435,16","454742,454742,16",
                                 "456725,456745,16","465830,465879,16","465901,465950,16",
                                 "484409,484410,16","490960,490979,16","492181,492182,16",
                                 "498824,498824,16");
    var abcRules     = new Array("560,651,16");
    
    var thisRules;
  
    switch(cardType)
    {
        case "A"  : thisRules = amexRules;
                    break;
        case "D"  : thisRules = discRules;
                    break;
        case "DC" : thisRules = dinersRules;
                    break;
        case "E"  : thisRules = enrouteRules;
                    break;
        case "J"  : thisRules = jcbRules;
                    break;
        case "M"  : thisRules = mcRules;
                    break;
        case "S"  : thisRules = switchRules;
                    break;
        case "SO" : thisRules = soloRules;
                    break;
        case "V"  : thisRules = visaRules;
                    break;
        case "MA" : thisRules = maestroRules;
                    break;
        case "DA" : thisRules = deltaRules;
                    break;
        case "AB" : thisRules = abcRules;
                    break;
        default   : thisRules = new Array();
                    break;
    }
  
    var ndx;
    var ruleDetails;

    for(ndx = 0 ; ndx < thisRules.length && !result ; ++ndx)
    {
        thisRule    = thisRules[ndx];
        ruleDetails = thisRule.split(",");
        
        var hiPrefix        = ruleDetails[0];
        var loPrefix        = ruleDetails[1];
        var valLengths      = ruleDetails[2].split("/");
        var cvcLength       = ruleDetails[3]
        var issueLength     = ruleDetails[4];
        var startDateLength = ruleDetails[5];
        
        /*
         * Verify card number length
         */
         
        var cardPrefix = cardNbr.substr(0,hiPrefix.length);
        if(cardPrefix >= hiPrefix && cardPrefix <= loPrefix)
        {
            if(valLengths[0] != 0)
            {
                result = false;
                for(lengthNdx = 0 ; lengthNdx < valLengths.length && !result ; ++lengthNdx)
                {
                    if(cardNbr.length == valLengths[lengthNdx])
                    {
                      result = true;
                    }
                }
            }
            else
            {
                result = true;
            }

            if(result)
            {
            }
            
            if(result)
            {
                cc_switch_startDateLength = startDateLength;
                cc_switch_issueLength     = issueLength;
                cc_cvc_length             = cvcLength;
            }
        }
    }
    
    return(result);
}

/*
 *****************************************************************************
 * Returns the name of the card type (for errors)
 *****************************************************************************
 */

function cardTypeName(cardType)
{
    var result;
  
    switch(cardType)
    {
        case "A" : result = "American Express";
                   break;
        case "D" : result = "Discover";
                   break;
        case "DC": result = "Diners Club";
                   break;
        case "E" : result = "Enroute"
                   break;
        case "J" : result = "JCB";
                   break;
        case "M" : result = "MasterCard";
                   break;
        case "S" : result = "Switch";
                   break;
        case "SO": result = "SOLO";
                   break;
        case "V" : result = "VISA";
                   break;
        case "MA": result = "Maestro";
                   break;
        case "DA": result = "Delta";
                   break;
        case "AB": result = "Australian BankCard";
                   break;
        default  : result = "unknown";
                   break;
    }
  
    return(result);
}
/*
 *****************************************************************************
 * Validate card verification code (CVV2/CID)
 *****************************************************************************
 */
 
function checkCvc(cardCvc)
{
    var result = true;
    
    if(cc_cvc_length > 0)
    {
        result = (cardCvc.length == cc_cvc_length);
    }
    
    return(result);
}

/*
 *****************************************************************************
 * Verify the card number
 *****************************************************************************
 * Incoming Parameters: 
 *   cardType:  A  - American Express
 *              AB - Australian BankCard
 *              D  - Discover
 *              DA - Delta
 *              DC - Diners Club
 *              E  - Enroute
 *              J  - JCB
 *              M  - MasterCard
 *              MA - Maestro
 *              S  - Switch
 *              SO - SOLO
 *              V  - VISA
 *   cardNbr:   Numeric card number, appropriate length determined by card 
 *              type/prefix.
 *
 */

function verifyCard(cardType,cardNbr)
{
    var result = true;
  
    if(cardNbr != "")
    {
        result = checkDigits(cardNbr);
        if(result == true)
        {
            result = checkPrefixAndLength(cardType,cardNbr);
            if(result == true)
            {
                result = luhnCheck(cardNbr);
                if(result == false)
                {
                    cc_fail_reason = "Incorrect checksum";
                }
            }
            else
            {
                cc_fail_reason = "Card length or type is incorrect";
            }
        }
        else
        {
            cc_fail_reason = "Card number must be fully numeric"; 
        }
  
        if (result == false)
        {
            alert("The " + cardTypeName(cardType) + " card number entered is incorrect!");
        }
    }
    else
      result = false;
  
    return(result);
}



src/tests/testcases/RandomCreditCardNumberGenerator.java
import java.util.List;
import java.util.Stack;
import java.util.Vector;

/**
 * See the license below. Obviously, this is not a Javascript credit card number
 * generator. However, The following class is a port of a Javascript credit card
 * number generator.
 * 
 * @author robweber
 * 
 */
public class RandomCreditCardNumberGenerator {
	/*
	 * Javascript credit card number generator Copyright (C) 2006 Graham King
	 * graham@darkcoding.net
	 * 
	 * This program is free software; you can redistribute it and/or modify it
	 * under the terms of the GNU General Public License as published by the
	 * Free Software Foundation; either version 2 of the License, or (at your
	 * option) any later version.
	 * 
	 * This program is distributed in the hope that it will be useful, but
	 * WITHOUT ANY WARRANTY; without even the implied warranty of
	 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
	 * Public License for more details.
	 * 
	 * You should have received a copy of the GNU General Public License along
	 * with this program; if not, write to the Free Software Foundation, Inc.,
	 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
	 * 
	 * www.darkcoding.net
	 */

	public static final String[] VISA_PREFIX_LIST = new String[] { "4539",
			"4556", "4916", "4532", "4929", "40240071", "4485", "4716", "4" };

	public static final String[] MASTERCARD_PREFIX_LIST = new String[] { "51",
			"52", "53", "54", "55" };

	public static final String[] AMEX_PREFIX_LIST = new String[] { "34", "37" };

	public static final String[] DISCOVER_PREFIX_LIST = new String[] { "6011" };

	public static final String[] DINERS_PREFIX_LIST = new String[] { "300",
			"301", "302", "303", "36", "38" };

	public static final String[] ENROUTE_PREFIX_LIST = new String[] { "2014",
			"2149" };

	public static final String[] JCB_16_PREFIX_LIST = new String[] { "3088",
			"3096", "3112", "3158", "3337", "3528" };

	public static final String[] JCB_15_PREFIX_LIST = new String[] { "2100",
			"1800" };

	public static final String[] VOYAGER_PREFIX_LIST = new String[] { "8699" };

	static String strrev(String str) {
		if (str == null)
			return "";
		String revstr = "";
		for (int i = str.length() - 1; i >= 0; i--) {
			revstr += str.charAt(i);
		}

		return revstr;
	}

	/*
	 * 'prefix' is the start of the CC number as a string, any number of digits.
	 * 'length' is the length of the CC number to generate. Typically 13 or 16
	 */
	static String completed_number(String prefix, int length) {

		String ccnumber = prefix;

		// generate digits

		while (ccnumber.length() < (length - 1)) {
			ccnumber += new Double(Math.floor(Math.random() * 10)).intValue();
		}

		// reverse number and convert to int

		String reversedCCnumberString = strrev(ccnumber);

		List<Integer> reversedCCnumberList = new Vector<Integer>();
		for (int i = 0; i < reversedCCnumberString.length(); i++) {
			reversedCCnumberList.add(new Integer(String
					.valueOf(reversedCCnumberString.charAt(i))));
		}

		// calculate sum

		int sum = 0;
		int pos = 0;

		Integer[] reversedCCnumber = reversedCCnumberList
				.toArray(new Integer[reversedCCnumberList.size()]);
		while (pos < length - 1) {

			int odd = reversedCCnumber[pos] * 2;
			if (odd > 9) {
				odd -= 9;
			}

			sum += odd;

			if (pos != (length - 2)) {
				sum += reversedCCnumber[pos + 1];
			}
			pos += 2;
		}

		// calculate check digit

		int checkdigit = new Double(
				((Math.floor(sum / 10) + 1) * 10 - sum) % 10).intValue();
		ccnumber += checkdigit;

		return ccnumber;

	}

	public static String[] credit_card_number(String[] prefixList, int length,
			int howMany) {

		Stack<String> result = new Stack<String>();
		for (int i = 0; i < howMany; i++) {
			int randomArrayIndex = (int) Math.floor(Math.random()
					* prefixList.length);
			String ccnumber = prefixList[randomArrayIndex];
			result.push(completed_number(ccnumber, length));
		}

		return result.toArray(new String[result.size()]);
	}

	public static String[] generateMasterCardNumbers(int howMany) {
		return credit_card_number(MASTERCARD_PREFIX_LIST, 16, howMany);
	}

	public static String generateMasterCardNumber() {
		return credit_card_number(MASTERCARD_PREFIX_LIST, 16, 1)[0];
	}

	public static boolean isValidCreditCardNumber(String creditCardNumber) {
		boolean isValid = false;

		try {
			String reversedNumber = new StringBuffer(creditCardNumber)
					.reverse().toString();
			int mod10Count = 0;
			for (int i = 0; i < reversedNumber.length(); i++) {
				int augend = Integer.parseInt(String.valueOf(reversedNumber
						.charAt(i)));
				if (((i + 1) % 2) == 0) {
					String productString = String.valueOf(augend * 2);
					augend = 0;
					for (int j = 0; j < productString.length(); j++) {
						augend += Integer.parseInt(String.valueOf(productString
								.charAt(j)));
					}
				}

				mod10Count += augend;
			}

			if ((mod10Count % 10) == 0) {
				isValid = true;
			}
		} catch (NumberFormatException e) {
		}

		return isValid;
	}

	public static void main(String[] args) {
		int howMany = 0;
		try {
			howMany = Integer.parseInt(args[0]);
		} catch (Exception e) {
			System.err
					.println("Usage error. You need to supply a numeric argument (ex: 500000)");
		}
		String[] creditcardnumbers = generateMasterCardNumbers(howMany);
		for (int i = 0; i < creditcardnumbers.length; i++) {
			System.out.println(creditcardnumbers[i]
					+ ":"
					+ (isValidCreditCardNumber(creditcardnumbers[i]) ? "valid"
							: "invalid"));
		}
	}
}

testcases/gencc.py
#!/usr/bin/python

"""
gencc: A simple program to generate credit card numbers that pass the MOD 10 check
(Luhn formula).
Usefull for testing e-commerce sites during development.

Copyright 2003 Graham King

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
"""

from random import Random
import sys
import copy

visaPrefixList = [ 	['4', '5', '3', '9'], 
	            	['4', '5', '5', '6'], 
               		['4', '9', '1', '6'],
                    ['4', '5', '3', '2'], 
             	    ['4', '9', '2', '9'],
                    ['4', '0', '2', '4', '0', '0', '7', '1'],
                    ['4', '4', '8', '6'],
                    ['4', '7', '1', '6'],
                    ['4'] ]

mastercardPrefixList = [    ['5', '1'],
                            ['5', '2'],
                            ['5', '3'],
                            ['5', '4'],
                            ['5', '5'] ]

amexPrefixList = [  ['3', '4'],
                    ['3', '7'] ]

discoverPrefixList = [ ['6', '0', '1', '1'] ]

dinersPrefixList = [    ['3', '0', '0'],
                        ['3', '0', '1'],
                        ['3', '0', '2'],
                        ['3', '0', '3'],
                        ['3', '6'],
                        ['3', '8'] ]

enRoutePrefixList = [   ['2', '0', '1', '4'],
                        ['2', '1', '4', '9'] ]

jcbPrefixList16 = [   ['3', '0', '8', '8'],
                    ['3', '0', '9', '6'],
                    ['3', '1', '1', '2'],
                    ['3', '1', '5', '8'],
                    ['3', '3', '3', '7'],
                    ['3', '5', '2', '8'] ]

jcbPrefixList15 = [ ['2', '1', '0', '0'],
                    ['1', '8', '0', '0'] ]

voyagerPrefixList = [ ['8', '6', '9', '9'] ]                    
                    

"""
'prefix' is the start of the CC number as a string, any number of digits.
'length' is the length of the CC number to generate. Typically 13 or 16
"""
def completed_number(prefix, length):

    ccnumber = prefix

    # generate digits

    while len(ccnumber) < (length - 1):
    	digit = generator.choice(['0',  '1', '2', '3', '4', '5', '6', '7', '8', '9'])
    	ccnumber.append(digit)


    # Calculate sum 

    sum = 0
    pos = 0

    reversedCCnumber = []
    reversedCCnumber.extend(ccnumber)
    reversedCCnumber.reverse()

    while pos < length - 1:

        odd = int( reversedCCnumber[pos] ) * 2
        if odd > 9:
            odd -= 9

        sum += odd

        if pos != (length - 2):

            sum += int( reversedCCnumber[pos+1] )

        pos += 2

    # Calculate check digit

    checkdigit = ((sum / 10 + 1) * 10 - sum) % 10

    ccnumber.append( str(checkdigit) )
    
    return ''.join(ccnumber)

def credit_card_number(generator, prefixList, length, howMany):

    result = []

    for i in range(howMany):
   
        ccnumber = copy.copy( generator.choice(prefixList) )

        result.append( completed_number(ccnumber, length) )

    return result

def output(title, numbers):

    result = []
    result.append(title)
    result.append( '-' * len(title) )
    result.append( '\n'.join(numbers) )
    result.append( '' )

    return '\n'.join(result)

#
# Main
#

generator = Random()
generator.seed()        # Seed from current time

print "darkcoding credit card generator "
print

mastercard = credit_card_number(generator, mastercardPrefixList, 16, 10)
print output("Mastercard", mastercard)

visa16 = credit_card_number(generator, visaPrefixList, 16, 10)
print output("VISA 16 digit", visa16)

visa13 = credit_card_number(generator, visaPrefixList, 13, 5)
print output("VISA 13 digit", visa13)

amex = credit_card_number(generator, amexPrefixList, 15, 5)
print output("American Express", amex)

# Minor cards

discover = credit_card_number(generator, discoverPrefixList, 16, 3)
print output("Discover", discover)

diners = credit_card_number(generator, dinersPrefixList, 14, 3)
print output("Diners Club / Carte Blanche", diners)

enRoute = credit_card_number(generator, enRoutePrefixList, 15, 3)
print output("enRoute", enRoute)

jcb15 = credit_card_number(generator, jcbPrefixList15, 15, 3)
print output("JCB 15 digit", jcb15)

jcb16 = credit_card_number(generator, jcbPrefixList16, 16, 3)
print output("JCB 16 digit", jcb16)

voyager = credit_card_number(generator, voyagerPrefixList, 15, 3)
print output("Voyager", voyager)



testcases/gencc.js
/*
Javascript credit card number generator 
Copyright (C) 2006 Graham King graham@darkcoding.net

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

www.darkcoding.net
*/

var visaPrefixList = new Array( 
    "4539",
    "4556",
    "4916",
    "4532",
    "4929",
    "40240071",
    "4485",
    "4716",
    "4" 
);

var mastercardPrefixList = new Array( 
    "51",
    "52",
    "53",
    "54",
    "55"
);

var amexPrefixList = new Array( 
    "34",
    "37"
);

var discoverPrefixList = new Array("6011");

var dinersPrefixList = new Array(
    "300",
    "301",
    "302",
    "303",
    "36",
    "38"
);

var enRoutePrefixList = new Array( 
    "2014",
    "2149"
);

var jcbPrefixList16 = new Array(
    "3088",
    "3096",
    "3112",
    "3158",
    "3337",
    "3528"
);

var jcbPrefixList15 = new Array(
    "2100",
    "1800"
);

var voyagerPrefixList = new Array("8699");


function strrev(str) {
   if (!str) return '';
   var revstr='';
   for (i = str.length-1; i>=0; i--)
       revstr+=str.charAt(i)
   return revstr;
}

/*
'prefix' is the start of the CC number as a string, any number of digits.
'length' is the length of the CC number to generate. Typically 13 or 16
*/
function completed_number(prefix, length) {

    var ccnumber = prefix;

    // generate digits

    while ( ccnumber.length < (length - 1) ) {
        ccnumber += Math.floor(Math.random()*10);
    }

    // reverse number and convert to int 

    var reversedCCnumberString = strrev( ccnumber );

    var reversedCCnumber = new Array();
    for ( var i=0; i < reversedCCnumberString.length; i++ ) {
        reversedCCnumber[i] = parseInt( reversedCCnumberString.charAt(i) );   
    }

    // calculate sum    
    
    var sum = 0;
    var pos = 0;
    
    while ( pos < length - 1 ) {

        odd = reversedCCnumber[ pos ] * 2;
        if ( odd > 9 ) {
            odd -= 9;
        }

        sum += odd;

        if ( pos != (length - 2) ) {

            sum += reversedCCnumber[ pos +1 ];
        }
        pos += 2;
    }
    
    // calculate check digit

    var checkdigit = (( Math.floor(sum/10) + 1) * 10 - sum) % 10;
    ccnumber += checkdigit;

    return ccnumber;
    
}

function credit_card_number(prefixList, length, howMany) {

    var result = new Array();
    for (var i = 0; i < howMany; i++) {

        var randomArrayIndex = Math.floor(Math.random() * prefixList.length); 
        var ccnumber = prefixList[ randomArrayIndex ];
        result.push( completed_number(ccnumber, length) );
    }

    return result;
}


